tap: introduce HTTP tap filter

mattklein123 opened the pull request:
This is a MVP for the HTTP tap filter. It includes minimal
infrastructure for the following:
1. Generic tap configuration which in the future will be used for
   static config, XDS config, etc. In this MVP the tap can be
   configured via a /tap admin endpoint.
2. Generic output configuration which in the future will be used for
   different output sinks such as files, gRPC API, etc. In this MVP
   the tap results are streamed back out the /tap admin endpoint.
3. Matching infrastructure. In this MVP only matching on request and
   response headers are implemented. Both logical AND and logical OR
   matches are possible.
4. In this MVP request/response body is not considered at all.
5. All docs are included and with all the caveats the filter is ready
   to use for the limited cases it supports (which are likely still to
   be useful).

There is a lot of follow on work which I will do in subsequent PRs.
This includes:
1. Merging the existing capture transport socket into this framework.
2. Implementing body support, both for matching on body contents as
   well as outputting body data.
3. Tap rate limiting so too many streams do not get tapped.
4. gRPC matching. Using reflection and loaded proto definitions, it will
   be possible to match on gRPC fields.
5. JSON matching. If the body parses as JSON, we can allow matching on
   JSON fields.

Part of https://github.com/envoyproxy/envoy/issues/1413.

*Risk Level*: Low. New feature with no dependencies. Minor changes to a few other files.
*Testing*: Unit/integration tests.
*Docs Changes*: Added.
*Release Notes*: Added.

mattklein123 commented:
@dio @lizan since this feature is of general interest to the Envoy/Istio community can one/both of you potentially own review of it? @htuch also in case you are interested. My plan is to refactor the transport socket capture feature to work within this framework next.

mattklein123 commented:
Replicating a comment from Slack:

@alyssawilk @zuercher @lizan With Lizan's help getting a log, I'm realizing that I don't think I can get the tap integration test working on OSX. The issue comes down to https://github.com/envoyproxy/envoy/issues/4294. Basically, the admin /tap response is a streamed response that never ends, and the test relies on the admin client being able to hang up when it's done getting data, but on OSX Envoy never sees the hangup and cleans up the admin client. Any thoughts here on what I can do? I can't think of any way to fix this so I'm wondering if I shouldn't run the test on OSX? Ideas? https://github.com/envoyproxy/envoy/pull/5515/files#diff-975313274dec5c7ffca88fb1be76d717R95

mattklein123 commented on source/extensions/filters/http/tap/tap_config_impl.cc:
```diff
@@ -0,0 +1,127 @@
+  request_headers->iterate(
+      [](const Http::HeaderEntry& header, void* context) -> Http::HeaderMap::Iterate {
+        envoy::data::tap::v2alpha::HttpBufferedTrace& trace =
+            *reinterpret_cast<envoy::data::tap::v2alpha::HttpBufferedTrace*>(context);
```
Originally this was for perf reasons when I was originally doing the 30K RPS benchmarks.

mattklein123 commented on source/extensions/common/tap/admin.cc:
```diff
@@ -0,0 +1,113 @@
+}
+
+void AdminHandler::submitBufferedTrace(std::shared_ptr<Protobuf::Message> trace) {
+  // TODO(mattklein123): The only reason this takes a shared_ptr is because currently a posted
```
Hmm yeah, bummer. I will remove this comment.

mattklein123 commented:
To follow up here, @lizan had the very good idea to have the test use HTTP/2 vs. HTTP/1 which will work. Unfortunately, admin is hard coded to use HTTP/1 currently, so I will make admin do auto codec detection. I will end up splitting all the admin changes out into a separate PR once I have this passing tests.

htuch commented:
@mattklein123 before diving into details, a few questions:
1. How far is this from https://docs.google.com/document/d/1fgVAH8BMrq_5dt8m54Rxp0OGXt6WlNUsAW0xGlBOkJs/edit#heading=h.hmtvuhdteovz? Is that still the reference design doc, or is it now obsoleted by new docs?
2. Small PRs; can we use this as a WiP/reference or for architecture discussion, and break into smaller PRs for review? I was pushing back on another 2k+ change earlier, and for consistency it might be good if maintainer authored PRs follow this advice. Is there a PR series that would make sense?

mattklein123 commented:
1) Yes that doc is still relevant and this change implements the MVP to support admin tapping.
2) Once I get this passing tests (which has been very annoying for various reasons), I'm going to break out all the ancillary changes into a separate PR which will be a few hundred lines changed. For the core MVP tap code, I don't see any useful way to break that up so I will leave that part as is.

mattklein123 commented:
@lizan @htuch ready for another pass. The bulk of this is docs/tests. I can't see any good way to split this up as a bunch of coverage I am relying on is from the integration test. If you have any ideas LMK otherwise we can stick with this.

mattklein123 commented on api/envoy/config/filter/http/tap/v2alpha/tap.proto:
```diff
@@ -0,0 +1,30 @@
+option java_package = "io.envoyproxy.envoy.config.filter.http.tap.v2alpha";
+option java_multiple_files = true;
+
+// [#protodoc-title: Tap]
```
I've been thinking of this as the noun tap as in wiretap. As such I would prefer to keep it this way unless people feel strongly otherwise.

htuch replied:
Fair enough, I was originally thinking TUN/TAP, but this is fine as long as it's consistent.

mattklein123 commented on source/extensions/filters/http/tap/tap_config_impl.cc:
```diff
@@ -0,0 +1,127 @@
+
+bool HttpPerRequestTapperImpl::onLog(const Http::HeaderMap* request_headers,
+                                     const Http::HeaderMap* response_headers) {
+  absl::optional<uint64_t> match_index;
```
I think this ends up being about the same number of lines of code as I already have, so I think I will keep it. :)

lizan replied:
I thought you may do `std::find(statuses_.begin(), statuses_.end(), HttpTapConfigImpl::MatchStatus{true, true})` though it is fine either.

lizan commented on source/extensions/filters/http/tap/tap_config_impl.cc:
```diff
@@ -0,0 +1,120 @@
+    const auto& status = statuses_[i];
+    if (status.request_matched_ && status.response_matched_) {
+      match_index = i;
+      break;
```
btw, is it guaranteed only one will match, or we only send the request to first config?

mattklein123 replied:
This is just checking for match specs that matched. It is possible that more than one matched if there are multiple OR conditions. Yes, in this case right now we just send the ID of the first one that matched. I suppose we could send an array of match IDs? Thoughts?

mattklein123 replied:
I think I'm going to get rid of this and go back to a single match condition, with the possibility of a complex set of AND/OR/NOT per discussion with @htuch so I'm going to resolve this.

htuch commented on api/envoy/service/tap/v2alpha/common.proto:
```diff
@@ -0,0 +1,88 @@
+  // [#comment:TODO(mattklein123): Rate limiting]
+}
+
+// Tap match configuration. Within a single match configuration, all conditions are evaluated as
```
Sure, that sounds great if you want to do that before finalizing the API; maybe either a TODO or tracking issue?

htuch commented on api/envoy/service/tap/v2alpha/common.proto:
```diff
@@ -0,0 +1,88 @@
+option java_multiple_files = true;
+
+// [#protodoc-title: Common tap configuration]
+
```
Ack

htuch commented on api/envoy/service/tap/v2alpha/common.proto:
```diff
@@ -0,0 +1,88 @@
+
+  // The tap output configuration. If a match configuration matches a data source being tapped,
+  // a tap will occur and the data will be written to the configured output.
+  OutputConfig output_config = 2 [(validate.rules).message.required = true];
```
What is the use case for multiple outputs? Won't it be just a file or the initiating RPC connection most of the time?

mattklein123 replied:
There are two things going on here, just to make sure we are on the same page.
1) There is a discrete `OutputConfig` message. The reason for this is that per the TODO eventually I would like to have some configuration over what gets tapped for perf reasons. For example, maybe only certain headers, maybe certain connection data, maybe only some number of body bytes.
2) In terms of multiple output sinks, I do have the repeated locked down to min/max 1 right now for implementation simplicity. I can see cases though where a user might want multiple outputs. For example, they are viewing admin stream but also want it written to file. Or, maybe they want a tap written to two different gRPC backends. I don't really feel that strongly about this, but it seemed better to future proof it with a repeated even if we set min/max to 1.

mattklein123 replied:
> One thing you could do is to make this a singleton for now and then upgrade to repeated later if you need; this is wire compatible friendly.

I don't think that's actually true when it comes to JSON/YAML, both in terms of singleton/repeated, as well as singular/plural. We would need to go through the deprecation cycle I think, which is why leaving it repeated w/ min/max set to 1 seems like a good future proof compromise. Though, like I said, I don't feel strongly about it so if you want to have a single sink I will change it.

htuch replied:
No, this is fine, I agree we don't want to introduce unnecessary churn to JSON/YAML. Field renaming is definitely bad, the whole list vs. singleton thing for JSON/YAML I keep forgetting; do you want to clarify this in https://github.com/envoyproxy/envoy/blob/master/api/STYLE.md?

mattklein123 replied:
Sure will do a separate PR.

htuch commented on api/envoy/admin/v2alpha/tap.proto:
```diff
@@ -0,0 +1,18 @@
+
+// The /tap admin request body that is used to configure an active tap session.
+message TapRequest {
+  // The opaque configuration ID used to match the configuration to a loaded extension.
```
The `TapRequest` can itself contain the full tap specification as long as there is  a single tap filter?

mattklein123 replied:
The request does take the full tap specification. Are you saying to not make the ID required and fail if there is ambiguity on which filter to target?

mattklein123 replied:
Think many listeners in the server, each of which has a tap filter installed. This is likely to be very common, for example, ingress and egress, etc.

htuch replied:
True, OK, this makes sense. The other alternative is to make this less opaque and allow specification of the particular listener, e.g. by listener index.

mattklein123 replied:
Yeah I suppose it could be by listener name (I prefer that over index since it's easier to understand), though this ID allows a name to span multiple listeners which IMO is kind of nice. I think I will keep this unless you think otherwise.

htuch replied:
Fair enough, just wanted to make sure all design points were covered.

htuch commented on docs/root/configuration/http_filters/tap_filter.rst:
```diff
@@ -0,0 +1,122 @@
+  very limited set of match conditions, output configuration, output sinks, etc. Capabilities will
+  be expanded over time and the configuration structures are likely to change.
+
+The HTTP tap filter is used to tap HTTP traffic. At a high level, the configuration is composed of
```
Writing nit: This first sentence is technically correct, but it might benefit from a more explanatory style, e.g. "The HTTP tap filter is used to interpose on and record HTTP traffic" or something like that.

htuch commented on docs/root/configuration/http_filters/tap_filter.rst:
```diff
@@ -0,0 +1,122 @@
+
+1. Match configuration: a list of conditions under which the filter will match an HTTP request
+   and begin a tap session.
+2. Output configuration: a list of output sinks that the filter will write the matched and tapped
```
Maybe link back to the RST for the API definitions for 1/2.

htuch commented on docs/root/configuration/http_filters/tap_filter.rst:
```diff
@@ -0,0 +1,122 @@
+2. Output configuration: a list of output sinks that the filter will write the matched and tapped
+   data to.
+
+Each of these concepts will be covered in the following example configuration section.
```
Maybe point out that this is a running example; it actually spans multiple sections to cover the concepts introduced above.

htuch commented on api/envoy/service/tap/v2alpha/common.proto:
```diff
@@ -0,0 +1,88 @@
+    //   configured from the :http:post:`/tap` admin endpoint. Thus, if an extension has
+    //   been configured to receive tap configuration from some other source (e.g., static
+    //   file, XDS, etc.) configuring the streaming admin output type will fail.
+    StreamingAdminSink streaming_admin = 1;
```
While I now grok the idea of potentially multiple sinks and a per-sink config, I wonder if we want this one to be more like `ResponseStreamSink`; this would apply in the gRPC case if the response comes back on bidi as well as the admin case.

mattklein123 replied:
Let me think about this. I think that could work. I guess the question is whether we would do a bidi for this. My thinking on how to handle gRPC is actually two different APIs: 1) an xDS for fetching tap configs, and 2) and access log / metrics service like API for pushing taps. This would allow these things to be decoupled. I.e., xDS could be used to configure a file sink, or a static config could be used to configure a gRPC push sink, for example.

htuch replied:
I can see that model working, but I also think that a common use case will be a client connecting to Envoy via gRPC. E.g. imagine I write a Wireshark-like client, I want to listen to some Envoy. I point my client at this Envoy and want it to stream stuff back to me. In the past whenever I've wanted to TAP, this has been the debug flow, rather than configuring a server endpoint to receive the TAP on. Network asymmetries could make that impractical.

I do get the long term direction you're heading here and the want to support a push as well as pull gRPC model, since the TAP might be used as part of a more sophisticated monitoring solution that is not under human debugger control, just pointing out that both models probably would be good to handle for gRPC as well.

mattklein123 replied:
Right that makes sense, but for the case you described, isn't this still over admin? Meaning, instead of a `ResponseStreamSink` wouldn't we actually allow the admin /tap endpoint to initiate a gRPC bidi stream, and then in the sink config we would have an output type which could be JSON/gRPC/etc.?

htuch replied:
Here's an example, based on a true story. You have the debug Wireshark-client under the control of a user in network X, Envoy is running in network Y. There are NAT/firewall between X/Y which allows X to initiate connections into Y but not vice versa. You need to be able to initiate all connections from the client in X.

mattklein123 replied:
We are on the same page. My point is that I think your use case is still an admin request/response since it's a direct connection into Envoy. I'm suggesting we would implement as options to the admin sink that describe whether it's JSON, gRPC, etc. Given that, does it make sense to change the name or leave it as `StreamingAdminSink` ?

htuch replied:
Yeah, we can leave it as is I guess. I actually think the gRPC connection would not be to the `/tap` endpoint, it would be to some proto specified service implied endpoint, but that's a more general "what should the admin gRPC interface look like?".

htuch commented on docs/root/configuration/http_filters/tap_filter.rst:
```diff
@@ -0,0 +1,122 @@
+``foo: bar`` is present OR a response header ``bar: baz`` is present. If either of these conditions
+are met, the request will be tapped and streamed out the admin endpoint. The difference between
+the first example and the second is the use of multiple top level match configurations to produce
+a logical OR.
```
This is a neat minimal use of structure to achieve AND/OR composition. What about NOT composition? Do we want to have a more general boolean predicate structure?

mattklein123 replied:
Yeah I thought about this some, and we might just want to do something similar to what we have done with RBAC which is to build AND/OR matching directly into the config structure. Let me think about this a bit more. I might want to do this in a self contained follow up since this PR stands alone and it's pretty big as is, and we can change the config since it's alpha. At minimum I will put in some TODOs. WDYT?

htuch replied:
Yes, fine for TODOs and the issue tracking matcher convergence.

mattklein123 replied:
I thought about this more and I'm going to rework this now to look more like RBAC, I think it's the right call. I will then reconcile RBAC/access log/etc. in subsequent work.

htuch commented on docs/root/configuration/http_filters/tap_filter.rst:
```diff
@@ -0,0 +1,122 @@
+
+The tap filter outputs statistics in the *http.<stat_prefix>.tap.* namespace. The :ref:`stat prefix
+<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.stat_prefix>`
+comes from the owning HTTP connection manager.
```
Would it be also useful to include the opaque ID here; even if strictly inferable, it seems a helpful hint.

mattklein123 replied:
I thought about this. I'm not sure if it's useful or not. I'm going to put in a TODO and we can see what kind of stats feedback we get.

htuch commented on docs/root/intro/version_history.rst:
```diff
@@ -13,6 +13,7 @@ Version history
 * http: added new grpc_http1_reverse_bridge filter for converting gRPC requests into HTTP/1.1 requests.
 * router: added ability to configure a :ref:`retry policy <envoy_api_msg_route.RetryPolicy>` at the
   virtual host level.
+* tap: added new :ref:`HTTP tap filter <config_http_filters_tap>`.
```
Draft?

htuch commented on source/extensions/common/tap/admin.cc:
```diff
@@ -0,0 +1,110 @@
+
+AdminHandler::AdminHandler(Server::Admin& admin, Event::Dispatcher& main_thread_dispatcher)
+    : admin_(admin), main_thread_dispatcher_(main_thread_dispatcher) {
+  bool rc =
```
NIt: const

htuch commented on source/extensions/common/tap/admin.cc:
```diff
@@ -0,0 +1,110 @@
+}
+
+AdminHandler::~AdminHandler() {
+  bool rc = admin_.removeHandler("/tap");
```
Ditto

htuch commented on source/extensions/common/tap/admin.cc:
```diff
@@ -0,0 +1,110 @@
+Http::Code AdminHandler::handler(absl::string_view, Http::HeaderMap&, Buffer::Instance& response,
+                                 Server::AdminStream& admin_stream) {
+  if (attached_request_.has_value()) {
+    response.add("An attached /tap admin stream already exists. Detach it.");
```
Is there a plan to support multiple streams? Seems like there's no need for this limit, even though it might make sense for the MVP.

mattklein123 replied:
Yeah I just did this for simplicity. I will add a TODO/comment here.

htuch commented on source/extensions/common/tap/admin.cc:
```diff
@@ -0,0 +1,110 @@
+  }
+}
+
+void AdminHandler::submitBufferedTrace(std::shared_ptr<Protobuf::Message> trace) {
```
Do you want stronger typing on the trace here? The code works with `Message`, but realistically we should be using a more specific type.

mattklein123 replied:
My thinking here is this is a generic interface that works regardless of the message type. When I port the capture transport socket over, this will allow a sink (admin, file, etc.) to implement a sink without knowing about the contained messages.

htuch replied:
I think for gRPC we're going to have to be more concrete though. Would we be sending `Any` objects out on the stream? Might be worth a few mins mocking up a gRPC service for this to see..

lizan replied:
Not sure using `Any` would help, if we send different types in different endpoint (gRPC method) then it is OK, otherwise I would define a message with oneof all possible types.

mattklein123 replied:
Yeah I like Lizan's idea of having a wrapper message with a oneof inside of it. Let me try that out.

htuch commented on source/extensions/common/tap/admin.h:
```diff
@@ -0,0 +1,76 @@
+    AttachedRequest(std::string config_id, Server::AdminStream* admin_stream)
+        : config_id_(std::move(config_id)), admin_stream_(admin_stream) {}
+
+    std::string config_id_;
```
Nit: const?

htuch commented on source/extensions/filters/http/tap/config.cc:
```diff
@@ -0,0 +1,46 @@
+Http::FilterFactoryCb TapFilterFactory::createFilterFactoryFromProtoTyped(
+    const envoy::config::filter::http::tap::v2alpha::Tap& proto_config,
+    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+
```
Nit: superfluous space

htuch commented on source/extensions/filters/http/tap/tap_config.h:
```diff
@@ -0,0 +1,78 @@
+  virtual void onResponseHeaders(const Http::HeaderMap& headers) PURE;
+
+  /**
+   * Called when the request is being destroyed and is being logged.
```
If this is really only called on destruction, maybe rename to `OnDestroyLog`?

htuch commented on source/extensions/filters/http/tap/tap_config.h:
```diff
@@ -0,0 +1,78 @@
+  /**
+   * @return a new per-request HTTP tapper which is used to handle tapping of a discrete request.
+   */
+  virtual HttpPerRequestTapperPtr newPerRequestTapper() PURE;
```
Nit: prefer `createPerRequestTrapper()` for consistency with other factory similar methods in code base.

htuch commented on source/extensions/filters/http/tap/tap_config_impl.cc:
```diff
@@ -0,0 +1,120 @@
+      admin_streamer_(admin_streamer) {
+  // TODO(mattklein123): The streaming admin output sink is the only currently supported sink.
+  ASSERT(admin_streamer);
+  ASSERT(proto_config_.output_config().sinks()[0].has_streaming_admin());
```
Should these be `RELEASE_ASSERT`s, sine they are subject to untrusted user input?

mattklein123 replied:
This is verified by schema. I think regular assert is OK?

htuch replied:
Fair enough, but this is a bit fragile, as when we extend the proto these might become weaker than expected without noticing. Since you're the one who owns this code right now, I'm not that worried, so we can resolve this.

htuch commented on source/extensions/filters/http/tap/tap_filter.cc:
```diff
@@ -0,0 +1,87 @@
+}
+
+const std::string& FilterConfigImpl::adminId() {
+  ASSERT(proto_config_.has_admin_config());
```
RELEASE_ASSERT?

mattklein123 replied:
Same, verified by schema?

mattklein123 commented:
@htuch @lizan I bit the bullet and rewrote the matching system to be a lot more generic and include and/or/any/not, etc. In doing so, I realized that the matching we need to do to support streaming matching (compute a match as data goes by vs. have everything available) is a lot more complicated than what we do for other matching in Envoy, and I'm skeptical we can share everything at a high level. I do think we can share the low-level matchers (headers, subnet, etc.) and I will work on that.

Anyway, PTAL. I added a bunch of comments on how the matchers work with regard to streaming. Even though this MVP does not support stream tapping today, we are setup to do so, and it's something that I know we will need to support shortly.

htuch commented on api/envoy/service/tap/v2alpha/common.proto:
```diff
@@ -0,0 +1,101 @@
+// configurations to be built using various logical operators.
+message MatchConfig {
+  // A set of match configurations used for logical operations.
+  message Set {
```
Nit: prefer `MatchSet` to indicate that everything in here implies a match condition.

htuch commented on api/envoy/service/tap/v2alpha/common.proto:
```diff
@@ -0,0 +1,101 @@
+
+// Tap match configuration. This is a recursive structure which allows complex nested match
+// configurations to be built using various logical operators.
+message MatchConfig {
```
Nit: prefer `MatchPredicate`

htuch commented on api/envoy/service/tap/v2alpha/common.proto:
```diff
@@ -0,0 +1,101 @@
+    // matches.
+    Set and_match = 2;
+
+    // A negation match. The match configuration will match if the negated match condition does not
```
Nit: double negative?

mattklein123 commented:
@htuch @lizan updated per comments

htuch commented on source/extensions/common/tap/tap.h:
```diff
@@ -0,0 +1,124 @@
+ * - The matching system is composed of a constant matching configuration. This is essentially
+ *   a tree of matchers given logical AND, OR, NOT, etc.
+ * - A per-stream/request matching status must be kept in order to compute interim match status.
+ * - In order to make this computationally efficient, the matching tree is kept in a vector, with
```
I think it would be a good idea to explain how the match tree is mapped to a vector here. Is this a root-leaf path, or maybe a heap structure or maybe a flattened N-ary tree, etc.?

mattklein123 replied:
I don't know the technical name. I guess it's a flattened N-ary tree. If you tell me the name I will update it or I will just call it a "flattened n-ary tree."

htuch replied:
I think the correct term would be "preorder traversal flattening of N-arary tree", I think you put the parent node down before the descendents, right?

mattklein123 replied:
OK. Can you approve and I will change this comment in a follow up? I have quite a few changes stacked up behind this one and I would rather not wait for another merge/CI round.

htuch commented on source/extensions/common/tap/tap.h:
```diff
@@ -0,0 +1,124 @@
+   * @param statuses supplies the per-stream-request match status vector which must be the same
+   *                 size as the match tree vector (see above).
+   */
+  virtual bool updateMatchStatus(const Http::HeaderMap* request_headers,
```
There's something a bit weird here to the casual reader about the mix of pure interface and concrete implementation. We have very low level details, e.g. tree representation, in this interface class while still leaving some methods PURE. How many plausible implementations are there of this interface given this?

mattklein123 replied:
There are quite a few implementations, but I will just move to tap_matcher.h with the concrete implementations.

htuch commented on docs/root/configuration/http_filters/tap_filter.rst:
```diff
@@ -0,0 +1,134 @@
+.. _config_http_filters_tap:
+
+Tap
```
Can we document here how streaming matching works? E.g. if you have a [request header, response header] match condition, and you've seen and matched only request headers, do you dump, then if you see a non-response header match you don't dump that?

mattklein123 replied:
Streaming tapping is not implemented yet for output. It is implemented in the sense that if you match on request headers, but there is no response, it will still dump the matching request headers at the end. I will clarify.

htuch commented on source/extensions/common/tap/tap_matcher.cc:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  // Per above, move the matcher into its position.
+  matchers[new_matcher->index()] = std::move(new_matcher);
```
How come we need the vector inlining? Can't we just build a regular tree and maintain a pointer to the current node of interest, tracking the leaf-parent path booleans as a vector if necessary?

mattklein123 replied:
Sorry I don't follow what you are suggesting. This implementation allows a per-request status vector to be created and then trivially updated without worrying about pointers. IMO it's both the simplest and most efficient implementation for what we need to do.

htuch replied:
Fair enough, I don't disagree that it's an efficient implementation. I'm visualizing this as at any point in time, for a given request, you have the match tree (immutable and shared) and your current status, which is those parts that have matched. I can see the attractiveness of the flattening due to the fact that you can easily map either way between match predicate and status. I

mattklein123 commented on test/extensions/common/tap/tap_matcher_test.cc:
```diff
@@ -0,0 +1,51 @@
+  envoy::service::tap::v2alpha::MatchPredicate config_;
+};
+
+TEST_F(TapMatcherTest, Any) {
```
Yeah, we should have targeted unit tests here. All of the other matchers have coverage via the integration tests, but will definitely add more UTs here in future PRs.
